docs/madr/decisions/078-special-mtp-algo.md

```diff
@@ -0,0 +1,118 @@
+# MeshTrafficPermission Is a Special Case of Inbound Policies and Requires a Custom Matching Algorithm
+
+* Status: accepted
+
+## Context and Problem Statement
+
+In the [Inbound Policies MADR](066-inbound-policies.md) we said all inbound policies are going to have similar structure:
+
+```yaml
+spec:
+  targetRef: {}
+  rules:
+     - matches: []
+       default: {}
+```
+
+and we'll create a unified matching algorithm that'll work for all inbound policies and potentially use Envoy Matching API.
+
+But turns out, it's a real challenge to come up with unified algorithm that'd work both for `MeshTrafficPermission` and other policies.
+
+Technically it's possible to build a matching tree for Envoy that'd be similar for all inbound policies.
+But, the size of the tree is getting big really fast.
+
+Let's look at the example:
+
+```yaml
+rules:
+   - matches:
+        - spiffeId:
+            type: Prefix
+            value: spiffe://trust-domain.mesh/
+     default: $conf1
+   - matches:
+        - path:
+            type: Prefix
+            value: /foo
+     default: $conf2
+   - matches:
+        - method: GET
+     default: $conf3
+```
+
+We’re matching on three independent dimensions:
+
+1. Whether the SPIFFE ID starts with `spiffe://trust-domain.mesh/`
+2. Whether the request path starts with `/foo`
+3. Whether the request method is `GET`
+
+Each condition has two possible outcomes: it either matches or doesn’t.
```

Comment:
The MADR overall looks good to me. 
Can we add some explanation that the above example is not from a single user-defined policy and it's actually a composite of all user-defined **policies** that elements under `rules` do not maintain a sequencial order?
It costed me some time to figure out why there can be 8 combinations because they look comimg from a single policy and we could just match the rules one by one.

Reply:
Single or composite is not actually relevant in this scenario. It'll be 8 combinations coming from a single policy. 

> we could just match the rules one by one

That's the issue with MeshTrafficPermission, we can't match only the first matching rule, we need to check all matching policies. And if we want to check all policies we need to build a tree for envoy that has 8 leaves

Reply:
Understood, thanks!
So this is decided by the different semantics between MTP and other policies.